docs/design/2022-04-21-performance-diagnosis-enhancements.md

```diff
@@ -0,0 +1,176 @@
+- Author(s): [cfzjywxk](https://github.com/cfzjywxk), [you06](https://github.com/you06)
+- Discussion PR:
+- Tracking Issue: #33963, #34106
+
+## Table of Contents
+
+* [Introduction](#introduction)
+* [Motivation or Background](#motivation-or-background)
+* [Detailed Design](#detailed-design)
+* [Impacts & Risks](#impacts--risks)
+
+## Introduction
+
+To enhance the diagnosis and tuning of TiDB, this RPC proposes some optimization of metrics, including the supplement of metrics, better display in Grafana.
+
+## Motivation or Background
+
+TiDB collects many metrics, but few of them are frequently used and the metrics of the critical paths are in different panels, which makes the diagnosis a problem. Furthermore, some latency events are not recorded in critical paths.
+
+This RPC proposes some improvements to the metrics collection, and display.
+
+## Detailed Design
+
+### D**ifferentiate Between Internal and External Requests**
+
+The tracking issue of it is #33963, and it should be applied to most of the metrics.
+
+#### Internal Transaction Usage
+
+The main goal of this task is to differentiate the internal and external transactions and mark the snapshots and transactions as internal for internal sessions and tasks.
+
+Some tasks are not running in the internal sessions, however, they are executed directly by `RunInNewTxn` :
+
+- DDL Backfill
+- Auto-ID management
+- Version check when bootstrapping
+- GC placement and label rules
+- Load data for cached tables
+- Alloc global ID for some operations
+    - Write sequence update to binlog
+    - Create temporary table
+
+When TiDB bootstrapping, it’ll create some internal sessions, these sessions are used for:
+
+- Bind info read/write
+- Privilege read/write
+- System variable read/write
+- Telemetry report
+- Statistic load
+
+The coprocessor request, transaction, and snapshot created by `RunInNewTxn` and the internal sessions need to be marked as internal.
+
+#### Session
+
+Add a flag in `SessionVars`, TiDB creates some internal sessions when bootstrap, and the internal flags are set after bootstrap.
+
+```diff
+type SessionVars struct {
+	...
++	Internal bool
+}
+```
+
+#### Coprocessor
+
+```diff
+// Request represents a kv request.
+type Request struct {
+	...
++	Internal bool
+}
+```
+
+#### KV Client
+
+Add internal flag in transaction and snapshot.
+
+```diff
+// KVTxn contains methods to interact with a TiKV transaction.
+type KVTxn struct {
+	...
++	Internal bool
+}
+```
+
+```diff
+// KVSnapshot implements the tidbkv.Snapshot interface.
+type KVSnapshot struct {
+	...
++	Internal bool
+}
+```
+
+#### kvproto
+
+```diff
+// Miscellaneous metadata attached to most requests.
+message Context {
+    reserved 4;
+    ...
++   // whether it's from internal request.
++   bool internal = xx;
```

Comment:
I'm thinking about adding an `enum` type to identify the `source type` of this request.
For example:
```
enum SrcType {
    DDLBackFillIndex = 01

    ...
    ExternalInsert = 10000
}
```

like this.

So we could do more detailed information.

Reply:
Looks good, I'll change it into enum type.

Reply:
If we tend to differentiate external insert and select, etc, it'll be a statement-level tag 🤔

Reply:
We have several different scopes like transaction, statement, and request levels. Maybe we could derive this flag from up to bottom?

Reply:
> derive this flag from up to bottom

So `SrcType` contains a prefix with "internal" or "external", and can be extended by some statement types, the read and lock requests have the statement types, and commit statements do not.

Statement-level types for the internal transactions may be skipped.

Reply:
Another solution is just using some simple integers, and internal and external types use different ranges.

Reply:
BTW `SrcType` is a bit confusing and not a good name.

Reply:
> Another solution is just using some simple integers

TiKV can report metrics with this tag, so a string can make the metrics display more readable.

Reply:
> BTW `SrcType` is a bit confusing and not a good name.

What about `request_source`?

Reply:
This looks good.